Spherical_harmonics functions

pilotchute opened the pull request:
Initial pull request for spherical.py and test_spherical.py
- [x] four tests passing on local machine
- [x] check codecov before merging

pilotchute commented on src/geocat/comp/spherical.py:
```diff
@@ -0,0 +1,85 @@
+
+    results = []
+    scale0 = 1 / (np.sum(input_scale, axis=(0, 1)) * ss.sph_harm(0, 0, 0, 0)**2)
+    # if input_scale is xr.DataArray:
+    #     scale0 = scale0.persist()
```
is this needed to prevent repeat computes of `scale0`?

pilotchute commented on src/geocat/comp/spherical.py:
```diff
@@ -0,0 +1,85 @@
+    #     scale0 = scale0.persist()
+    scale1 = scale0 * 2
+    input_data_scaled = np.multiply(input_data, input_scale)
+    # if input_data is xr.DataArray:
+    #     input_data_scaled = input_data_scaled.persist()
```
is this needed to prevent repeat computes of `input_data_scaled`?

pilotchute commented:
found a performance issue that was forcing a large .compute() call. have found a fix, but it's a large change, and I'm going to change this to a draft until the bugs are gone.

pilotchute commented:
fixed now, and I'm much happier with the way the computation is structured.

erogluorhan commented:
Is this still a draft or ready for review once the build failures and stalls are fixed?

pilotchute commented:
It's almost ready, I found a function in scipy last night that could calculate the voronoi area of points on a sphere, so I added a helper function to take theta and phi arrays, turn them into an array of x,y,z points to input to the `scipy.spatial.SphericalVoronoi(points((n,3)).calculate_areas()` function. 
Also added tests for those. 
I think it's about ready.
I'm not having it calculate the voronoi area by default due to the performance of the scipy function being quite slow, (and I don't want to have it calcuated every harmonic run, when if it's passed in, it only has to be calculated once and passed in any number of times. 
if scipy ever gets their voronoi math to meet `n*log(n)` speed of the fastest published area calculation algs I'll reconsider, but for now scipy is using a simpler `n^2` implementation.

pilotchute commented:
I still need to finish the docstrings, and give people a chance to read the code and see if I missed any sanity checks, or have anything that could cause problems in pathological cases.

erogluorhan commented:
> I still need to finish the docstrings, and give people a chance to read the code and see if I missed any sanity checks, or have anything that could cause problems in pathological cases.

All sound good! Just ping us when it is ready.

pilotchute commented:
@erogluorhan should I put in an optional parameter to instruct the decomposition to also provide an array of the `[m,n]` values for the output harmonics?

erogluorhan commented on src/geocat/comp/spherical.py:
```diff
@@ -0,0 +1,231 @@
+        the spherical harmonic decomposition of the input data
+    """
+
+    scale_val = 1 / (np.sum(scale, axis=(0, 1)) *
```
A description of this scale var would be helpful

pilotchute replied:
Sure thing, it's not simple, but it's easy to explain. (the math to figure it out took me a while and a few sheets of paper.)

erogluorhan commented on test/test_spherical.py:
```diff
@@ -0,0 +1,85 @@
+test_results_xr = xr.DataArray(test_results_np, dims=['har']).compute()
+
+
+def test_decomposition_np():
```
Not a requirement but for getting these test cases directly debuggable in PyCharm (i.e. right click in the code and debug), you may consider putting your test case funcs into a class; something like:

```
from unittest import TestCase

class Test_spherical(TestCase):
    def test_decomposition_np(self):
    def test_...
```

pilotchute replied:
I like this, I'll add it.

pilotchute replied:
done

erogluorhan commented on src/geocat/comp/spherical.py:
```diff
@@ -0,0 +1,231 @@
+
+    mlist = []
+    nlist = []
+    scale_mul = []
```
Describing what is done for `mlist` and `nlist` in these loops would be helpful, too. The same applies to `recomposition`, too.

pilotchute replied:
Sure, will add some comments.

pilotchute replied:
done

erogluorhan commented on src/geocat/comp/spherical.py:
```diff
@@ -0,0 +1,231 @@
+        scale_dat = np.expand_dims(np.multiply(data, scale), axis=(2))
+
+    # if xarray, set dims and chunks for broadcast in ss.sphere_harm
+    if type(data) is xr.DataArray:
```
Since SupportedTypes = Union[np.ndarray, xr.DataArray] already, I'd expect and if-else block instead of two if's. The same applies to `recomposition`, too.

pilotchute replied:
two ifs leaves open future support for additional data types. if else would be harder to expand. and there is no performance difference. 

willing to if-else if you feel strongly about it.

erogluorhan replied:
No strong feelings; just wondered the intention, and it makes sense.

erogluorhan commented on src/geocat/comp/spherical.py:
```diff
@@ -0,0 +1,231 @@
+        theta = xr.DataArray(theta, dims=data.dims).chunk((chunk_size))
+        phi = xr.DataArray(phi, dims=data.dims).chunk((chunk_size))
+
+    results = np.sum(
```
I am curious if using `np.sum` and `np.multiply` would be a performance-cutter for chunked Xarrays (i.e. Dask data arrays). @anissa111 , any thoughts?

pilotchute replied:
my testing showed that this was the fastest implementation. (and I did a lot of testiing) they are dask chunk compatible and I an show dask localcluster utilization graphs.

pilotchute replied:
![image](https://user-images.githubusercontent.com/20648306/158468267-3891810f-f931-43ca-8ed0-ce644b441f56.png)

pilotchute replied:
That shows chunked test data creation, decomposition, and recomposition.

erogluorhan replied:
I see. I asked this as I remembered @anissa111 had an issue in relhum/dewtemp when she first used `np` functions, then she had to switch to using dask.array.functions. 

Did you have a chance to compare this current version vs. `dask.array.sum` and `dask.array.multiply`?

pilotchute replied:
they also don't require separate computation calls in an if/else statement to avoid breaking on numpy.

erogluorhan replied:
> they also don't require separate computation calls in an if/else statement to avoid breaking on numpy.

Yeah, I agree with this making the code nicer

erogluorhan reviewed (commented):
The implementation here looks great! I have a few inline comments, though!

erogluorhan commented:
> @erogluorhan should I put in an optional parameter to instruct the decomposition to also provide an array of the `[m,n]` values for the output harmonics?

This sounds like a good enhancement to the function, though, I don't know of any use case

erogluorhan commented on src/geocat/comp/spherical.py:
```diff
@@ -0,0 +1,231 @@
+    Returns
+    -------
+    decomposition : :class:`numpy.ndarray`, :class:`xarray.DataArray`
+        the spherical harmonic decomposition of the input data
```
It might be helpful to put a "See also" part that explains which NCL functions can be handled with these functions here (just as we did in climatologies.py and some other places)

erogluorhan commented on src/geocat/comp/spherical.py:
```diff
@@ -0,0 +1,231 @@
+        The size of the each edge of the dask chunks if using xarray.DataArray inputs.
+        Some arrays will be 2d, and others 1d, and the final calculation operates on a 3d array.
+        thus the chunks used in the largest calculation scale at chunk_size^3
+        A chunk size of 256 is reccomended. Defaults to 'auto'
```
typo "recommended"

pilotchute replied:
done

pilotchute commented:
> > @erogluorhan should I put in an optional parameter to instruct the decomposition to also provide an array of the `[m,n]` values for the output harmonics?
> 
> This sounds like a good enhancement to the function, though, I don't know of any use case

I could see a case where a user might want to know the shape a particular result's harmonic. The only reason I'm wary of implementing it is I don't think it's a great idea to have a function provide incompatible output  based on an input option. 

I think I would instead prefer to have a separate function to return the set of harmonics that the end user could then compare result(n) harm(n) to get the harmonic components.

erogluorhan commented:
> I think I would instead prefer to have a separate function to return the set of harmonics that the end user could then compare result(n) harm(n) to get the harmonic components.

Yes, this makes sense to me!

anissa111 reviewed (commented):
This looks good! I have a couple very minor changes, but then I'll approve!

erogluorhan commented:
I approved but noticed later that some CI tests are failing. When they are fixed, I'll approve.

pilotchute commented on test/test_spherical.py:
```diff
@@ -0,0 +1,150 @@
+import xarray as xr
+import sys
+import unittest
+from typing import Union
```
done

pilotchute commented:
@erogluorhan the CI failures are due to conda not finding an environment solution in python 3.7. it isn't due to the merge, and we have discussed not supporting python 3.7 in the 2022.03.0 release. 
Should we not worry about the CI failures on 3.7?

erogluorhan commented:
> @erogluorhan the CI failures are due to conda not finding an environment solution in python 3.7. it isn't due to the merge, and we have discussed not supporting python 3.7 in the 2022.03.0 release. Should we not worry about the CI failures on 3.7?

Ok, no problem then! Approved!

erogluorhan reviewed (approved):
Great work!

anissa111 reviewed (approved):
Woo!

pilotchute commented:
@erogluorhan this will need to be force-merged with your privileges, since the 3.7 CI is a requirement for a normal merge.
